### PR TITLE
ddl: fix issue of add foreign key failed but doesn't rollback

### LINKED ISSUE #39582: issue of add foreign key failed but doesn't rollback
## Reproduce

```sql
 set @@global.tidb_enable_foreign_key=1;
set @@foreign_key_checks=1;
create table t1 (id int key,a int, index(a));
create table t2 (id int key,a int);
insert into t1 values (1,1);
```

```sql
test> ALTER TABLE t1 ADD foreign key fk(a) references t2(id) ON DELETE CASCADE;
(1452, 'Cannot add or update a child row: a foreign key constraint fails (`test`.`t1`, CONSTRAINT `fk` FOREIGN KEY (`a`) REFERENCES `t2` (`id`) ON DELETE CASCADE)')
test> show create table t1\G
***************************[ 1. row ]***************************
Table        | t1
Create Table | CREATE TABLE `t1` (
  `id` int(11) NOT NULL,
  `a` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  KEY `a` (`a`),
  CONSTRAINT `fk` FOREIGN KEY (`a`) REFERENCES `test`.`t2` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

Add foreign key failed, but `show create table` result still contain the foreign key info.

## TiDB Version
```sql
***************************[ 1. row ]***************************
tidb_version() | Release Version: v6.5.0-alpha-248-g6b4739f170
Edition: Community
Git Commit Hash: 6b4739f170a1fb53a10d3f4bd50ad6288460f252
Git Branch: master
UTC Build Time: 2022-12-02 06:31:34
GoVersion: go1.19.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore
```

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Issue Number: close #39582

Problem Summary: Fix issue of add foreign key failed but doesn't rollback

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

**Review by lcwangchao — APPROVED** [review]

File: ddl/foreign_key.go [hunk 1]
```diff
@@ -94,54 +97,46 @@ func onDropForeignKey(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ err
...
+	if job.IsRollingback() {
+		job.FinishTableJob(model.JobStateRollbackDone, model.StateNone, ver, tblInfo)
+	} else {
 		job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
-		job.SchemaState = fkInfo.State
-		return ver, nil
-	default:
-		return ver, dbterror.ErrInvalidDDLState.GenWithStackByArgs("foreign key", fkInfo.State)
 	}
+	job.SchemaState = model.StateNone
```
Comment (tangenta) [inline ddl/foreign_key.go:138]: Why is it `StateNone`?

File: ddl/foreign_key_test.go [hunk 2]
```diff
@@ -1513,6 +1513,24 @@ func TestAddForeignKey(t *testing.T) {
 	tbl2Info = getTableInfo(t, dom, "test", "t2")
 	require.Equal(t, 0, len(tbl2Info.ForeignKeys))
 	tk.MustGetDBError("alter table t2 drop index idx_c, add constraint fk_c foreign key (c) references t1(b)", dbterror.ErrDropIndexNeededInForeignKey)
+
+	// Test cyclical Dependencies add foreign key failed.
```
Comment (tangenta) [inline ddl/foreign_key_test.go:1517]:
```suggestion
	// Test circular dependency add foreign key failed.
```

File: ddl/foreign_key.go (hunk 1 above)
Comment (crazycs520) [reply ddl/foreign_key.go:138]: For drop column/index/foreign_key job, the final `job.SchemaState` are both `StateNone`

**Review by tangenta — APPROVED** [review]

**Pull request merged**